otel: simple OTEL collector/Jaeger/Aspire stack for testing purposes

krissetto opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**
Added a compose stack to be used for testing out OTEL functionality

**- How I did it**
- Added compose stack;
- Added basic OTEL collector config;
- Added basic usage doc

**- How to verify it**
Read the doc and try it out :)

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog.
It must be placed inside the below triple backticks section:
-->
```markdown changelog
Add simple compose file for testing OTEL traces locally
```

**- A picture of a cute animal (not mandatory but encouraged)**

![](https://s-media-cache-ak0.pinimg.com/736x/23/e1/67/23e167ec47369e2da74fe2c886853f87.jpg)

thaJeztah commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+  aspire-dashboard:
+    image: mcr.microsoft.com/dotnet/nightly/aspire-dashboard
+    ports:
+      - 18888:18888
```
wondering if we should use `127.0.0.1` for these; just in case someone runs this without considering that mapping the port will default to the public interface. @akerouanton any thoughts?

thaJeztah replied:
If we're updating; do you think it's better to be explicit (`0.0.0.0:18888:18888`) for these? Might give more of a clue that it's listening to "any" IP (not sure if we changed semantics on IPv4/IPv6 for that, but I think we still consider `0.0.0.0` to be "either IPv4/6")

krissetto replied:
made this more explicit

thaJeztah commented on hack/otel/usage.md:
```diff
@@ -0,0 +1,28 @@
+
+## How can I use it?
+
+1) Export the env var used to override the OTLP endpoint:  
```
Nit; should this be a actual markdown numbered list?

krissetto replied:
Mmh good question, I thought it looked a bit better but I guess it depends on the markdown renderer used. No hard feelings here, if it's "more correct" to do otherwise i'll change this

krissetto replied:
sounds good to me! I'll fix up this pr today since it's been lingering for a while

thaJeztah replied:
I very much forgot it was still there 😂 ❤️

akerouanton commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+services:
+
+  jaeger:
+    image: jaegertracing/all-in-one:latest
```
I think it's worth adding `restart: always` to every service.

krissetto replied:
Good idea. Reading the compose spec i see `always: The policy always restarts the container until its removal.`. 
Maybe using `unless-stopped` might be better in this case? I wouldn't want potential users (even if we kinda only expect devs and power-users to use this) to accidentally have these services running forever

akerouanton replied:
I thought `unless-stopped` wouldn't be right here because AFAIK if the container is stopped during dockerd shutdown, it won't automatically restart (eg. on a reboot). But I tried it and it seems to work properly, so we can also consider using that restart policy.

akerouanton replied:
Could you address this please?

akerouanton commented on hack/otel/otelcol.yaml:
```diff
@@ -0,0 +1,21 @@
+receivers:
+  otlp:
+    protocols:
+      grpc:
+      http:
```
Just pulled `otel/opentelemetry-collector:latest` and it now defaults to `localhost`, so we need this instead:

```suggestion
      grpc:
        endpoint: 0.0.0.0:4317
      http:
        endpoint: 0.0.0.0:4318
```

krissetto commented:
I addressed the feedback given here, @akerouanton if you can confirm it still works for your workflow after the port mapping and otelcol changes that'd be great

akerouanton commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+      DOTNET_DASHBOARD_UNSECURED_ALLOW_ANONYMOUS: 'true'
+  
+  otelcol:
+    image: otel/opentelemetry-collector:latest
```
The `otel/opentelemetry-collector` image only contains a minimal set of processors, connectors, and no filters. So it's not particularly useful beyond just forwarding to multiple exporters.

Could you switch to `otel/opentelemetry-collector-contrib` instead? This would allow us to easily experiment with processing pipelines directly from this repo.

```suggestion
    image: otel/opentelemetry-collector-contrib:latest
```

akerouanton commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+      - aspire-dashboard
+    ports:
+      - 4318:4318 # default otlp http port
+    volumes:
+      # Mount the otelcol.yml config file
+      - ./otelcol.yaml:/etc/otelcol/config.yaml
```
For the same reason (ease experimentation from this repo), could you switch to a Compose watch?

```suggestion
    develop:
      watch:
        - action: sync+restart
          path: ./otelcol.yaml
          target: /etc/otelcol-contrib/config.yaml
```

akerouanton commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+    ports:
+      - 16686:16686
+
+  aspire-dashboard:
```
Could you add `restart: always` to this service?

akerouanton commented on hack/otel/compose.yaml:
```diff
@@ -0,0 +1,27 @@
+  otelcol:
+    image: otel/opentelemetry-collector:latest
+    restart: always
+    depends_on:
+      - jaeger
+      - aspire-dashboard
```
AFAIK you can't manually import traces into Aspire, but you can with Jaeger. This is useful for instance when you want to load the OTel traces we dump at the end of integration tests on the CI. (I used that trick a couple of times to debug DNS-related issues on the CI)

OTOH, Aspire supports OTel logs and metrics. The Engine doesn't emit them, but this is useful if you want to experiment with a processing pipeline that transforms traces into metrics.

I'm planning to add a pipeline or two to have an easy way to track the performance of various operations in libnetwork (e.g. `libnetwork.drivers.bridge.createNetwork` and `daemon.connectToNetwork`) to better understand the impact of some of the changes we're making (e.g. `iptables` vs `nftables`).

Jaeger is also more "barebone", so I tend to use it for looking at traces, and then switching to Aspire to double-check that my pipelines are correctly exporting metrics. I could do everything with Aspire, so it's mostly personal preference.

Long story short, it's definitely useful to have both, and I don't think it'd add much value to have a Compose profile to choose which one to run.

krissetto replied:
i can include the profiles if we want that, and will implement the suggestions above as well!

thaJeztah replied:
Thanks! Perhaps in that case we don't need the profiles. I was mostly curious if we needed both, or if we could pick "the best one". Looks like there's still some parts not provided by both, so probably fine to keep'em both for now.

thaJeztah replied:
Slightly wondering if `contrib/` would be a more appropriate place for this compose-file 🤔

tianon replied:
> Slightly wondering if `contrib/` would be a more appropriate place for this compose-file 🤔

100% agree (hack is traditionally for scripts related to building, this is more of an example, which is what contrib is for)

krissetto commented:
@thaJeztah could you add the `impact/` label so the CI isn't angry? or is it best to remove the changelog msg? it's a bit confusing

thaJeztah commented:
Added the label; we can remove it later if we don't think it's worth including in the release notes 👍

thaJeztah commented on contrib/otel/usage.md:
We can probably rename this to `README.md` then it shows up on Github if someone browses to the directory 😄

krissetto replied:
done

krissetto committed:
Basic compose file for testing OTEL bits

Signed-off-by: Christopher Petito <chrisjpetito@gmail.com>

akerouanton commented:
I think we don't need a changelog entry for this one as it's not a user-facing change.

thaJeztah reviewed (approved):
LGTM
